File: nodejs/src/handlers/rest.js
'use strict';

const { generate } = require('../payload/generator');
const { handlePayloadError } = require('./errors');

function mimeForFormat(format) {
  switch ((format || 'json').toLowerCase()) {
    case 'xml': return 'application/xml';
    case 'text': return 'text/plain';
    case 'html': return 'text/html';
    default: return 'application/json';
  }
}

function writePayloadResponse(res, size, format, method, resource) {
  try {
    const payload = generate(size, format, method, resource);
    const bytes = Buffer.byteLength(payload, 'utf8');
    res.set({
      'Content-Type': mimeForFormat(format),
      'X-Mock-Size-KB': String(size),
      'X-Mock-Format': (format || 'json').toLowerCase(),
      'X-Mock-Method': method,
      'X-Mock-Resource': resource,
      'X-Payload-Bytes': String(bytes),
    });
    res.status(200).send(payload);
  } catch (err) {
    if (!handlePayloadError(res, err)) throw err;
  }
}

function parseAPIPath(path) {
  const p = (path || '').replace(/^\//, '');
  if (!p) return null;
  const parts = p.split('/').filter(Boolean);
  if (parts.length === 1) return { resource: parts[0], id: null };
  if (parts.length === 2) return { resource: parts[0], id: parts[1] };
  return null;
}

function restHandler(req, res) {
  const parsed = parseAPIPath(req.path);
  if (!parsed) {
    res.status(404).end();
    return;
  }
  const size = parseInt(req.query.size, 10) || 1;
  const format = req.query.format || 'json';
  const { resource, id } = parsed;

  switch (req.method) {
    case 'GET':
      writePayloadResponse(res, size, format, 'GET', id ? `${resource}/${id}` : resource);
      break;
    case 'POST':
      if (id) res.status(404).end();
      else writePayloadResponse(res, size, format, 'POST', resource);
      break;
    case 'PUT':
      if (!id) res.status(404).end();
      else writePayloadResponse(res, size, format, 'PUT', `${resource}/${id}`);
      break;
    case 'DELETE':
      if (!id) res.status(404).end();
      else writePayloadResponse(res, size, format, 'DELETE', `${resource}/${id}`);
      break;
    default:
      res.status(405).send('Method Not Allowed');
  }
}

module.exports = { restHandler };

File: go/internal/handler/rest.go
package handler

import (
	"net/http"
	"strconv"
	"strings"

	"github.com/SidathWeerasinghe/mock-services/go/internal/payload"
)

func mimeForFormat(format string) string {
	switch strings.ToLower(format) {
	case "xml":
		return "application/xml"
	case "text":
		return "text/plain"
	case "html":
		return "text/html"
	default:
		return "application/json"
	}
}

func writePayloadResponse(w http.ResponseWriter, size int, format, method, resource string) {
	payloadStr, err := payload.Generate(size, format, method, resource)
	if handlePayloadError(w, err) {
		return
	}
	w.Header().Set("Content-Type", mimeForFormat(format))
	w.Header().Set("X-Mock-Size-KB", strconv.Itoa(size))
	w.Header().Set("X-Mock-Format", strings.ToLower(format))
	w.Header().Set("X-Mock-Method", method)
	w.Header().Set("X-Mock-Resource", resource)
	w.Header().Set("X-Payload-Bytes", strconv.Itoa(len([]byte(payloadStr))))
	w.WriteHeader(http.StatusOK)
	_, _ = w.Write([]byte(payloadStr))
}

func parseSizeFormat(r *http.Request) (size int, format string) {
	size = 1
	format = "json"
	if s := r.URL.Query().Get("size"); s != "" {
		if v, err := strconv.Atoi(s); err == nil {
			size = v
		}
	}
	if f := r.URL.Query().Get("format"); f != "" {
		format = f
	}
	return size, format
}

func parseAPIPath(path string) (resource, id string, ok bool) {
	path = strings.TrimPrefix(path, "/api/")
	if path == "" || strings.HasPrefix(path, "/") {
		return "", "", false
	}
	parts := strings.Split(path, "/")
	if len(parts) == 1 {
		return parts[0], "", true
	}
	if len(parts) == 2 {
		return parts[0], parts[1], true
	}
	return "", "", false
}

// REST dispatches /api/* requests by method.
func REST(w http.ResponseWriter, r *http.Request) {
	resource, id, ok := parseAPIPath(r.URL.Path)
	if !ok {
		http.NotFound(w, r)
		return
	}
	size, format := parseSizeFormat(r)
	switch r.Method {
	case http.MethodGet:
		if id != "" {
			writePayloadResponse(w, size, format, "GET", resource+"/"+id)
		} else {
			writePayloadResponse(w, size, format, "GET", resource)
		}
	case http.MethodPost:
		if id == "" {
			writePayloadResponse(w, size, format, "POST", resource)
		} else {
			http.NotFound(w, r)
		}
	case http.MethodPut:
		if id != "" {
			writePayloadResponse(w, size, format, "PUT", resource+"/"+id)
		} else {
			http.NotFound(w, r)
		}
	case http.MethodDelete:
		if id != "" {
			writePayloadResponse(w, size, format, "DELETE", resource+"/"+id)
		} else {
			http.NotFound(w, r)
		}
	default:
		http.Error(w, "method not allowed", http.StatusMethodNotAllowed)
	}
}
